### PR TITLE
Rework PR/MR feature

### PR DESCRIPTION
**Pull request by neomilium**
This one is bigger than previous ones... even I tried to fine split.

I worked in TDD, so I started with tests, then made some progress, saw that my work will affect more than expected, wrote more tests, continue with a light refactoring and finally rework all PR/MR as a generic `GitService`.

The hardest part on this topic is to "implement" all unspecified behavior, so I tried to extend the test coverage.
I tried to keep the backward compatibility, but some specific corner cases could fail.

At this point, I made my best to prevent from breaking compatibility but I think some fundamentals should be re-written to make sense.

Nevertheless this rewrite allow us to easily fix some issues later, and provide a generic way to use "git service", which can be reuse for other features (e.g. #158).

### CODE REVIEW

**Review by bastelfreak — APPROVED** [review]
Looks good to me, but it doesn't hurt if someone reviews this with more Ruby knowledge. @ekohl @raphink can one of you take a look as well?

**Review by smortex — APPROVED** [review]
Lots of stuff, not easy for a review, but overall LGTM!

**Comment by neomilium** [comment]
I spotted a bug, and will fix it tomorrow.

**Comment by neomilium** [comment]
@smortex @bastelfreak @bittner : Its ready to review!

File: lib/modulesync/git_service.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module ModuleSync
+  # Namespace for Git service classes (ie. GitHub, GitLab)
+  module GitService
+    def self.instantiate(type:, options:)
```
Comment (ekohl) [inline lib/modulesync/git_service.rb]: This is really a factory. Should the class be named `GitServiceFactory` or is that too Java-y?

**Review by ekohl — COMMENTED** [review]
I had a quick look through the code and in general it looks good to me. It's pretty large so I may have missed things.

File: spec/unit/modulesync/pr/github_spec.rb [hunk 2]
```diff
@@ -22,7 +22,7 @@
     it 'submits PR when --pr is set' do
       allow(@client).to receive(:pull_requests).with(@git_repo, :state => 'open', :base => 'master', :head => "#{@namespace}:#{@options[:branch]}").and_return([])
       expect(@client).to receive(:create_pull_request).with(@git_repo, 'master', @options[:branch], @options[:pr_title], @options[:message]).and_return({"html_url" => "http://example.com/pulls/22"})
-      expect { @it.manage(@namespace, @repo_name, @options) }.to output(/Submitted PR/).to_stdout
+      expect { @it.open_pull_request(@namespace, @repo_name, @options) }.to output(/Submitted PR/).to_stdout
```
[the remark this answers is not in the document]
Comment (neomilium) [reply spec/unit/modulesync/pr/github_spec.rb]: Agree, I tried to minimize code alteration so I follow existing syntax, but it should definitely use a more modern syntax, without globals.
Opening a new issue could be a way to keep the note?

File: lib/modulesync/git_service/base.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  module GitService
+    # Generic class for git services
+    class Base
+      def open_pull_request(repo_path:, namespace:, title:, message:, source_branch:, target_branch:, labels:, noop:) # rubocop:disable Metrics/ParameterLists, Metrics/LineLength
+        unless source_branch != target_branch
+          raise ModuleSync::Error,
+                "Unable to open a pull request with the same source and target branch: '#{source_branch}'"
+        end
+
+        _open_pull_request(
```
[the remark this answers is not in the document]
Comment (neomilium) [reply lib/modulesync/git_service/base.rb:11]: Make sense.

File: features/update/pull_request.feature [hunk 4]
```diff
@@ -9,8 +9,9 @@ Feature: Create a pull-request/merge-request after update
         github: {}
       """
     And I set the environment variables to:
-      | variable     | value  |
-      | GITHUB_TOKEN | foobar |
+      | variable        | value                    |
+      | GITHUB_TOKEN    | foobar                   |
+      | GITHUB_BASE_URL | https://github.faker.com |
```
[the remark this answers is not in the document]
Comment (neomilium) [reply features/update/pull_request.feature:14]: OK, thanks.

**Comment by neomilium** [comment]
@ekohl Codecov report is not up-to-date: I just removed a commit that adds `cucumber` to tests because it disturb the code coverage. I opened an issue here #236.

**Review by smortex — COMMENTED** [review]
One minor thing

**Review by smortex — APPROVED** [review]

**Comment by bastelfreak** [comment]
I think it's good enough to be merged now
